src/Connection/FiniteConnectionPool.php

```diff
@@ -0,0 +1,336 @@
+<?php
+
+namespace Amp\Http\Client\Connection;
+
+use Amp\CancellationToken;
+use Amp\Deferred;
+use Amp\Http\Client\Internal\ForbidSerialization;
+use Amp\Http\Client\Request;
+use Amp\Http\Client\Response;
+use Amp\Promise;
+use Amp\Success;
+use function Amp\call;
+use function Amp\coroutine;
+
+final class FiniteConnectionPool implements ConnectionPool
+{
+    use ForbidSerialization;
+
+    /** @var int */
+    private $maxHttp1Connections;
+
+    /** @var int */
+    private $maxHttp2Connections;
+
+    /** @var ConnectionFactory */
+    private $connectionFactory;
+
+    /** @var Promise[][] */
+    private $connections = [];
+
+    /** @var Deferred[][] */
+    private $waiting = [];
+
+    /** @var bool[] */
+    private $waitForPriorConnection = [];
+
+    /** @var int */
+    private $totalConnectionAttempts = 0;
+
+    /** @var int */
+    private $totalStreamRequests = 0;
+
+    /** @var int */
+    private $openConnectionCount = 0;
+
+    /** @var callable */
+    private $requestToKeyMapper;
+
+    /**
+     * Create a connection pool that limits the number of connections per authority.
+     *
+     * @param int                    $maxHttp1Connections Maximum number of HTTP/2 connections allowed to a single authority.
+     * @param int                    $maxHttp2Connections Maximum number of HTTP/1.x connections allowed to a single authority.
```

Comment:
descriptions are swapped in the file :)

Reply:
Oops, and then I copied and pasted it everywhere! 😝 Fixed.